File: Leetcode100/leetcode100_python/MergekSortedLists.py
from typing import List

class ListNode:
    def __init__(self, val = 0, next = None):
        self.val = val
        self.next = next

class Solution:
    def mergeTwoLists(self, l1, l2):
        head = ListNode()
        p = head
        while l1 and l2:
            if l1.val < l2.val:
                p.next = l1
                l1 = l1.next
            else:
                p.next = l2
                l2 = l2.next
            p = p.next
        p.next = l1 if l1 else l2
        return head.next
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        def merge(l, r):
            if l == r:
                return lists[l]
            if l > r:
                return None
            mid = (l + r) // 2
            return self.mergeTwoLists(merge(l, mid), merge(mid + 1, r))
        return merge(0, len(lists) - 1)

if __name__ == "__main__":
    inputs = []
    line = input()
    while line:
        inputs.append([int(num) for num in line.split()])
        line = input()
    
    lists = []
    for nums in inputs:
        l = ListNode()
        p = l
        for num in nums:
            p.next = ListNode(num)
            p = p.next
        lists.append(l.next)

    solution = Solution()
    result = solution.mergeKLists(lists)
    while result:
        print(result.val, " ")
        result = result.next

File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/merge_k_sorted_lists/MergekSortedLists.java
package main.java.com.jiang.leetcode100.merge_k_sorted_lists;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) {this.val = val;}
    ListNode(int val, ListNode next) {this.val = val; this.next = next;}
}

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {

        ListNode head = new ListNode();
        ListNode p = head;

        while(true) {
            int min_val = Integer.MAX_VALUE;
            int index = -1;
            for(int i = 0; i < lists.length; i++) {
                if(lists[i] != null && lists[i].val < min_val) {
                    min_val = lists[i].val;
                    index = i;
                }
            }
            if(index == -1) break;
            p.next = lists[index];
            p = p.next;
            lists[index] = lists[index].next;
        }
        p.next = null;

        
        return head.next;
    }
}

public class MergekSortedLists {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        List<String> inputs = new ArrayList<>();
        while(!scanner.hasNext("#")) {
            inputs.add(scanner.nextLine());
        }
        scanner.close();
        
        ListNode[] lists = new ListNode[inputs.size()];
        for(int i = 0; i < inputs.size(); i++) {
            ListNode l = new ListNode();
            ListNode p = l;
            String[] nums_str = inputs.get(i).split(" ");
            for(int j = 0; j < nums_str.length; j++) {
                p.next = new ListNode(Integer.parseInt(nums_str[j]));
                p = p.next;
            }
            p.next = null;
            l = l.next;
            lists[i] = l;
        }

        ListNode result = new Solution().mergeKLists(lists);
        while(result != null) {
            System.out.println(result.val);
            result = result.next;
        }

        // for(int i = 0; i < lists.length; i++) {
        //     ListNode p = lists[i];
        //     while(p != null) {
        //         System.out.print(p.val + " ");
        //         p = p.next;
        //     }
        //     System.out.println("");
        // }
    }
}